charmcraft/commands/build.py

```diff
@@ -0,0 +1,178 @@
+"""
+
+# The minimum set of hooks to be provided for compatibility with old Juju
+HOOK_NAMES = [
+    'install',
+    'start',
+    'upgrade-charm',
+]
+
+
+def polite_exec(cmd):
+    """Execute a command, only showing output if error."""
+    p = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+
+    stdout = []
+    for line in p.stdout:
+        stdout.append(line)
+    retcode = p.wait()
+
+    if retcode:
+        logger.error("Execution ended in %d for cmd %s", retcode, cmd)
+        for line in stdout:
+            logger.debug(":: %s", line.rstrip())
+    return retcode
+
+
+def link(srcpath, destdir):
+    """Link a file."""
+    if not destdir.exists():
+        os.makedirs(destdir)
+
+    # XXX Facundo 2020-05-19: we could make hard links here, with the benefit of `juju deploy`
+    # being happy about these links, while at the same time getting the benefit of "changing
+    # a file in the project and being ready to deploy" (however, TRICKY!); that said, we would
+    # need to handle directories by hand (but currently we're not linking dirs...)
+    destpath = destdir / srcpath.name
+    destpath.symlink_to(srcpath)
+    return destpath
+
+
+def _build(charmdir, entrypoint):
+    """Main entry point."""
+    buildpath = charmdir / BUILD_DIRNAME
+    logger.debug("Building charm in %r", str(buildpath))
+
+    # XXX Facundo 2020-05-18: for now we're re-building always from scratch, in the
+    # future we *may* reuse stuff already inside
+    if buildpath.exists():
+        shutil.rmtree(str(buildpath))
+    buildpath.mkdir()
+
+    # basic files
+    logger.debug("Linking in basic files and charm code")
+    link(charmdir / CHARM_METADATA, buildpath)
+
+    # the charm code
+    # XXX Facundo 2020-05-18: for now, one file, may we copy the entrypoint and its whole dir/tree
+    # (if not charmdir); also we need to understand if we want to ignore some files (e.g. .pyc)
+    linked_entrypoint = link(entrypoint, buildpath / 'src')
+
+    # dispatch mechanism
+    logger.debug("Creating the dispatch mechanism")
+    dispatch_content = DISPATCH.format(
+        entrypoint_relative_path=linked_entrypoint.relative_to(buildpath))
+    dispatch_path = buildpath / DISPATCH_FILENAME
+    with open(dispatch_path, "wt", encoding="utf8") as fh:
+        fh.write(dispatch_content)
+    os.chmod(dispatch_path, dispatch_path.stat().st_mode | stat.S_IXUSR)
+
+    # bunch of symlinks, to support old juju
+    hookpath = buildpath / 'hooks'
+    hookpath.mkdir()
+    for hookname in HOOK_NAMES:
+        (hookpath / hookname).symlink_to(dispatch_path)
+
+    # dependencies
+    logger.debug("Installing dependencies")
+    # XXX Facundo 2020-05-18: we may want to be flexible with how to include the
+    # dependencies, e.g. respecting current lib directory, or not having a requirements file
+    libpath = buildpath / 'lib'
+    reqspath = charmdir / 'requirements.txt'
```

Comment:
I feel like we can be explicit about 'charmcraft.yaml', but everything else could be dependent on that. 1 file *must* be accessible to get started.